pkg/controller/applicationlayer/applicationlayer_controller.go

```diff
@@ -0,0 +1,319 @@
+type ReconcileApplicationLayer struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client          client.Client
+	scheme          *runtime.Scheme
+	provider        operatorv1.Provider
+	status          status.StatusManager
+	clusterDomain   string
+	licenseAPIReady *utils.ReadyFlag
+}
+
+// Reconcile reads that state of the cluster for a ApplicationLayer object and makes changes
+// based on the state read and what is in the ApplicationLayer.Spec
+func (r *ReconcileApplicationLayer) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ApplicationLayer")
+
+	applicationLayer, err := getApplicationLayer(ctx, r.client)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			reqLogger.Info("ApplicationLayer object not found")
+			r.status.OnCRNotFound()
+			return reconcile.Result{}, nil
+		}
+		r.status.SetDegraded("Error querying for Application Layer", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	preDefaultPatchFrom := client.MergeFrom(applicationLayer.DeepCopy())
+	// Set defaults
+	updateApplicationLayerWithDefaults(applicationLayer)
+
+	// Validate the configuration
+	if err := validateApplicationLayer(applicationLayer); err != nil {
+		r.status.SetDegraded("Invalid applicationLayer provided", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	// Write the application layer back to the datastore, so the controllers depending on this can reconcile.
+	if err = r.client.Patch(ctx, applicationLayer, preDefaultPatchFrom); err != nil {
+		log.Error(err, "Failed to write defaults")
+		r.status.SetDegraded("Failed to write defaults", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	variant, installation, err := utils.GetInstallation(ctx, r.client)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.status.SetDegraded("Installation not found", err.Error())
+			return reconcile.Result{}, err
+		}
+		r.status.SetDegraded("Error querying installation", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	pullSecrets, err := utils.GetNetworkingPullSecrets(installation, r.client)
+
+	if err != nil {
+		r.status.SetDegraded("Error retrieving pull secrets", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	lcSpec := applicationLayer.Spec.LogCollection
+
+	// if ApplicationLayer spec exists then LogCollection should be set
+	// TODO: when we will have multiple features in future this should change to at least one feature being set
+	if lcSpec == nil {
+		r.status.SetDegraded(fmt.Sprintf("Error retrieving LogCollection Spec"), "")
+		return reconcile.Result{}, err
```

Comment:
err here has not been set to anything useful, and will likely just be nil. I would suggest not return an error here, just log (please add the error logline) and set degraded (as you have).

When an error is returned from Reconcile, Reconcile is immediately retried. This is not a temporary error, this is a misconfiguration of a resource that is being watched. Until the user fixes the resource, this will fail, and when the user fixes the resource, the controller is notified because it is watching said resource. This means that returning an error here doesn't help.